emacs/merlin-cap.el

```diff
@@ -7,67 +7,484 @@
+(defun merlin-cap--exit-function (string _state)
+  "Print a message for completion STRING."
+  (let ((ret (merlin-cap--annotate string)))
+    (if ret (message "%s%s" string ret))))
 
 (defun merlin-cap--annotate (candidate)
-  "Retrieve the annotation for candidate CANDIDATE in
-`merlin-completion-annotate-table'."
-  (cdr (assoc candidate merlin-cap--table)))
+  "Retrieve the annotation for candidate CANDIDATE."
+  ;; Fancy completion styles (like partial-completion) can concatenate multiple candidates
+  ;; together into a single candidate (e.g. "List." and "map" concatenated into "List.map"),
+  ;; and each component of that will have its own annotation text property.
+  ;; We want to print the type of the overall expression, so grab the annotation from the end.
+  (get-text-property (1- (length candidate)) 'merlin-cap--annotation candidate))
+
+(defvar merlin-cap--interrupt-symbol nil)
+
+(define-error 'merlin-cap--test-interrupt "Test-only interrupt")
+(defun merlin-cap--interrupt-in-test (pos-symbol)
+  "Error if POSITION-SYMBOL is equal to `merlin-cap--interrupt-symbol'."
+  (when (eq pos-symbol merlin-cap--interrupt-symbol)
+    (signal 'merlin-cap--test-interrupt pos-symbol)))
+
+(defvar-local merlin-cap--process-last-event nil
+  "The most recent process event for a Merlin process in this buffer.")
+
+(defun merlin-cap--sentinel (process event)
+  "Store EVENT in `merlin-cap--process-last-event' in PROCESS's buffer."
+  (let ((buf (process-buffer process)))
+    (when (buffer-live-p buf)
+      (with-current-buffer buf
+        (setq merlin-cap--process-last-event event)))))
+
+;; Buffer-local variables for request buffers
+(defvar-local merlin-cap--request-debug-cons nil
+  "This request's entry in the variable `merlin-debug-last-commands'.
+
+This is a cons whose cdr starts out nil, is set to the symbol
+`pending' when the request is fully created, and is set to the
+raw output string from the request when it's completed.")
+(defvar-local merlin-cap--request-stderr-proc nil
+  "The stderr pipe process for this request.")
+(defvar-local merlin-cap--request-originating-buffer nil
+  "The buffer that this completion request is for.")
+(defvar-local merlin-cap--request-position nil
+  "The POSITION argument to `merlin-cap--complete-prefix' for this request.")
+(defvar-local merlin-cap--request-prefix nil
+  "The PREFIX argument to `merlin-cap--complete-prefix' for this request.")
+(defvar-local merlin-cap--request-ignore-region nil
+  "The IGNORE-REGION argument to `merlin-cap--complete-prefix' for this request.")
+
+(defun merlin-cap--pending-request (position ignore-region)
+  "Return a pre-existing request for POSITION and IGNORE-REGION.
+
+Specifically, return a cons whose car is the prefix that the
+request was for and whose cdr is the request buffer.  Return nil
+if no pending request exists, or its inputs do not match."
+  (let ((buf (current-buffer)))
+    (with-current-buffer (get-buffer-create " *merlin-async-proc-stdout*" t)
```

Comment:
`get-buffer-create` changed the number of arguments between 27 [*]() and 28 [**]() which is causing linting issues.

[1]: https://github.com/emacs-mirror/emacs/blob/emacs-27.2/src/buffer.c#L509-L518
[2]: https://github.com/emacs-mirror/emacs/blob/emacs-28.2/src/buffer.c#L532-L546